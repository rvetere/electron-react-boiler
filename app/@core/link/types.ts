import { IButtonProps } from '@core'

export interface ILinkProps extends IButtonProps {
  link: string
  text: string
  target?: string
  rel?: string
  button?: boolean
  inverted?: boolean
  fullSize?: boolean
  onMouseDown?: () => void
  onMouseMove?: (event: React.MouseEvent) => void
}
