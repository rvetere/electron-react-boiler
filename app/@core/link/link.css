.iconPadding svg {
  margin-right: $spacingSmall;
}

.link {
  display: inline-block;
  text-decoration: none;
  padding: 0;
  border: 0;
  background: 0;
  color: $linkColor;
  text-align: left;
  cursor: pointer;

  &:hover {
    color: $linkHoverColor;
  }

  &:active {
    color: $black;
  }

  &:focus {
    text-decoration: underline;
  }
}

.inverted {
  color: $whiteAlpha7;

  &:active,
  &:hover,
  &:focus {
    color: $white;
  }
}

.fullSizeOverlay {
  position: absolute;
  z-index: $zIndexNavigation;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  overflow: hidden;
  opacity: 0;
  background-color: $overlayColor;
  transition: all $transitionSpeedFast ease;

  &:hover {
    opacity: 1;
  }

  &:hover,
  &:focus {
    text-decoration: none;
  }
}
