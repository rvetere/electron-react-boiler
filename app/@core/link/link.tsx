import { getButtonStylings } from '@core'
import classNames from 'class-names'
import React, { Children, isValidElement, PureComponent, ReactNode } from 'react'
import { Link } from 'react-router-dom'
import styles from './link.css'
import { ILinkProps } from './types'

export class LinkComponent extends PureComponent<ILinkProps> {
  public render(): ReactNode {
    let { children, link, button, inverted, fullSize, text, ...otherProps } = this.props

    const count = Children.count(children)
    const iconPadding = count > 1 && (children && isValidElement(children[0]))
    let linkClasses

    if (button) {
      // get button classes and remove the button specific props
      const buttonStylings = getButtonStylings(children, otherProps)
      linkClasses = buttonStylings.className
      otherProps = buttonStylings.otherProps
    } else {
      linkClasses = classNames(styles.link, {
        [styles.inverted]: inverted,
        [styles.fullSizeOverlay]: fullSize,
        [styles.iconPadding]: iconPadding
      })
    }

    return (
      <Link to={link} className={linkClasses}>
        {text}
      </Link>
    )
  }
}
