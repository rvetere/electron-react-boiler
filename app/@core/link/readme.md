# Links

You can pass either a string with an url or an ApiRoute object (see linkFragment.graphql).
When passing an ApiRoute we're generating the effective route with the same logic as in the monolith to guarantee a fallback site in case it's not yet isomorphic.

If you use text in combination with an Icon spacings are automatically added.

**Never use an `<a>` tag directly, always use the Link component so we can change the routing logic in one place**

## Url as string

```javascript
<Link link={'/it'}>
  <Translation>More job vacancies</Translation>
</Link>
```

## Url as object

```javascript
<Link
  link={{
    sectorId: 's1',
    type: 'type',
    actionName: 'actionname',
    actionValue: { id: 'id', name: 'name' },
    parameters: [{ key: 'key', value: 'value' }],
    anchor: 'anchor'
  }}
>
  <Translation>More job vacancies</Translation>
</Link>

// -> '/de/s1/type/actionname/name-id?key=value#anchor'
```

## Link to route that isn't isomorphic yet

```javascript
<Link external link={'/'}>
  This Link leads to a non isomorphic page!
</Link>
```

## Prefetch

The JS for the route gets fetched once the current route was loaded and the Brwoser goes in idle mode.
**This only takes effect in production mode!**

```javascript
<Link prefetch link={'/site'}>
  <Translation>More job vacancies</Translation>
</Link>
```

## Replace instead of Push

In some cases you might want to replace the url instead of pushing a new route in the history

```javascript
<Link replace link={'/it'}>
  <Cart />
</Link>
```

## External links

```javascript
<Link target="_blank" link={'www.mysite.com'}>
  <Cart />
</Link>
```

## Fullsize

When using fullsize links make sure to include a description where the link is leading due to accessibility reasons.

```javascript
<Link fullSize link={'/it'}>
  <span className="ssrOnly">
    <Translation>Show location details</Translation>
  </span>
</Link>
```

## With Button stylings

For more possibilites check out the Button readme

```javascript
<Link button primary link={'/it'}>
  <Cart />
  <Translation>More job vacancies</Translation>
</Link>
```
