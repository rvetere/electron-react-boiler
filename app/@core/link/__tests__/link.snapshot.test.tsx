import React from 'react'
import renderer from 'react-test-renderer'
import { Link } from '../link'

describe('Link', () => {
  test('Link component should render correctly with absoluteUrl', async () => {
    const link = {
      absoluteUrl: '/some/action',
      parameters: []
    }
    const linkText = 'someText'
    const app = renderer.create(<Link link={link}>{linkText}</Link>)
    expect(app.toJSON()).toMatchSnapshot()
  })

  test('Link component should render correctly with route properties', () => {
    const link = {
      absoluteUrl: null,
      type: 'Sale',
      actionName: 'Default',
      actionValue: {
        id: 'someId',
        name: 'someName'
      },
      parameters: [
        {
          key: 'someKey',
          value: 'someValue'
        },
        {
          key: 'someOtherKey',
          value: 'someOtherValue'
        }
      ],
      anchor: null,
      sectorId: null
    }
    const linkText = 'someText'
    const app = renderer.create(<Link link={link}>{linkText}</Link>)

    expect(app.toJSON()).toMatchSnapshot()
  })
})
