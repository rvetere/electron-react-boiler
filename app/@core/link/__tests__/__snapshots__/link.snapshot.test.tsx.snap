// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link Link component should render correctly with absoluteUrl 1`] = `
<a
  className="link"
  href="/some/action"
  onClick={[Function]}
>
  someText
</a>
`;

exports[`Link Link component should render correctly with route properties 1`] = `
<a
  className="link"
  href="/sale/somename-someid?someKey=someValue&someOtherKey=someOtherValue"
  onClick={[Function]}
>
  someText
</a>
`;
